### PR TITLE
response outline for JSON responses

### PR DESCRIPTION
**Pull request by harshlele**
i've been using the app for a while, but one of the things ive always wanted is a outline for very long JSON responses, similar to how VS code does it, so i decided to see if i make it myself lol. I'd love it if you guys merged this. Ive added a screenshot: 

image

---

- [x] Better language detection (show outline without clicking in editor)
- [x] Close drop down on click (broken now)
- [x] `@mouseleave` don't work on mobile devices

**Deploy preview:** https://deploy-preview-1484--hoppscotch.netlify.app

### CODE REVIEW

**Comment by liyasthomas** [comment]
@harshlele this feature looks very handy. We're considering this PR for review.

**Comment by harshlele** [comment]
Thanks :)

>

File: components/ui/ace-editor.vue [hunk 1]
```diff
@@ -103,6 +176,7 @@ export default {
 
     this.editor = editor
     this.cacheValue = this.value
+    this.outline.init(this.value)
```
Comment (AndrewBastin) [inline components/ui/ace-editor.vue:179]: Ace Editor component is used for every type of request data. Not just JSON.

This makes the editor crash if the content is not JSON.

File: components/ui/ace-editor.vue [hunk 2]
```diff
@@ -111,6 +185,15 @@ export default {
       if (this.lint) this.provideLinting(content)
     })
```
Comment (AndrewBastin) [inline components/ui/ace-editor.vue:187]: We have an editable JSON editor for the POST request JSON bodies. Can we get live updates for that ?
(Make sure to debounce)

**Review by AndrewBastin — CHANGES_REQUESTED** [review]

File: components/ui/ace-editor.vue (hunk 2 above)
Comment (harshlele) [reply components/ui/ace-editor.vue:187]: so you basically want outlining for the post request editor as well?

Comment (AndrewBastin) [reply components/ui/ace-editor.vue:187]: The component used on the POST request editor is the same as the component used on the Ace Editor, just that it is writable. So the JSON path thingy already comes on top when it is edited.

**Comment by harshlele** [comment]
oof sorry if its messed up, not sure whats happened i had a lot of conflicts while rebasing on main

**Comment by liyasthomas** [comment]
You dont have to rebase untill the feature is complete.

**Comment by harshlele** [comment]
ahhhh, i thought that would remove the files, really sorry for making a mess out of this 😓

**Comment by AndrewBastin** [comment]
@harshlele any luck on resolving the issues I have raised ? I want to take a shot at solving both of those issues myself. But I am kinda on a break from the project due to a health issue.

**Comment by harshlele** [comment]
I have solved some of the things in the todo list! i was gonna push that
after i fixed the package.json issue but i couldn't do that :/ (and was
busy with work all day yesterday). but yeah if you want to do it feel free!

File: components/ui/ace-editor.vue (hunk 2 above)
Comment (harshlele) [reply components/ui/ace-editor.vue:194]: ive made it so that it updates the current path as you type now, the dropdown isnt yet visible though

**Comment by liyasthomas** [comment]
@harshlele I've removed `package.json` and `package-lock.json` from this PR.

**Comment by harshlele** [comment]
Thank you! 😀

File: components/ui/ace-editor.vue (hunk 2 above)
Comment (harshlele) [reply components/ui/ace-editor.vue:194]: ok now it should work in the request editor too! @AndrewBastin

File: components/ui/ace-editor.vue (hunk 1 above)
Comment (AndrewBastin) [reply components/ui/ace-editor.vue]: This seems to be fixed in the new commits. Resolving this convo.

File: components/ui/ace-editor.vue (hunk 2 above)
Comment (AndrewBastin) [reply components/ui/ace-editor.vue:194]: It works, but it kinda messes up some times. I guess we will disable the outline for the request body.

**Comment by AndrewBastin** [comment]
Okay, with the code to restrict currently the JSON outline to only the JSON Response Renderer. I am fine with this PR to be merged.

@harshlele Do you have anything else to add upon this before this is merged ?

File: components/ui/ace-editor.vue (hunk 2 above)
Comment (harshlele) [reply components/ui/ace-editor.vue:201]: messes up in what sense? for me the only time it doesnt work is when the json itself is invalid, but cant really do anything about that. did you notice any other issues?

Comment (AndrewBastin) [reply components/ui/ace-editor.vue:201]: When I tried it out, it struggled a bit to display the correct outline info sometimes, it lags behind sometimes.

**Comment by liyasthomas** [comment]
@harshlele if there's nothing else to add to this PR, we can merge this.

@AndrewBastin

**Review by AndrewBastin — APPROVED** [review]

**Comment by harshlele** [comment]
Yeah, merge it :)

**Pull request merged**